File: 2. Structural Design Patterns/Facade/Java/FacadeExample.java
// Facade Pattern Example in Java
// Subsystems: CPU, Memory, HardDrive; Facade: Computer

class CPU {
    void freeze(){ System.out.println("CPU: freeze"); }
    void jump(long position){ System.out.println("CPU: jump to " + position); }
    void execute(){ System.out.println("CPU: execute"); }
}
class Memory { void load(long position, byte[] data){ System.out.println("Memory: load " + data.length + " bytes at " + position); } }
class HardDrive { byte[] read(long lba, int size){ System.out.println("HardDrive: read " + size + " bytes from LBA " + lba); return new byte[size]; } }

class Computer {
    private final CPU cpu = new CPU();
    private final Memory memory = new Memory();
    private final HardDrive hardDrive = new HardDrive();
    void start(){ cpu.freeze(); byte[] boot = hardDrive.read(0, 1024); memory.load(0, boot); cpu.jump(0); cpu.execute(); }
    void shutdown(){ System.out.println("Computer: shutdown sequence"); }
}

public class FacadeExample {
    public static void main(String[] args){ Computer c = new Computer(); c.start(); c.shutdown(); }
}

File: 2. Structural Design Patterns/Facade/C#/FacadeExample.cs
// Facade Pattern Example in C#
// Subsystems: CPU, Memory, HardDrive; Facade: Computer

using System;

namespace FacadeExample
{
    class CPU { public void Freeze(){ Console.WriteLine("CPU: freeze"); } public void Jump(long position){ Console.WriteLine($"CPU: jump to {position}"); } public void Execute(){ Console.WriteLine("CPU: execute"); } }
    class Memory { public void Load(long position, byte[] data){ Console.WriteLine($"Memory: load {data.Length} bytes at {position}"); } }
    class HardDrive { public byte[] Read(long lba, int size){ Console.WriteLine($"HardDrive: read {size} bytes from LBA {lba}"); return new byte[size]; } }

    class Computer
    {
        private readonly CPU cpu = new();
        private readonly Memory memory = new();
        private readonly HardDrive hardDrive = new();
        public void Start(){ cpu.Freeze(); var boot = hardDrive.Read(0, 1024); memory.Load(0, boot); cpu.Jump(0); cpu.Execute(); }
        public void Shutdown(){ Console.WriteLine("Computer: shutdown sequence"); }
    }

    class Program
    {
        static void Main(){ var computer = new Computer(); computer.Start(); computer.Shutdown(); }
    }
}
